modules/services/window-managers/bspwm/default.nix

```diff
@@ -0,0 +1,77 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  cfg = config.xsession.windowManager.bspwm;
+  bspwm = cfg.package;
+
+  camelToSnake = s:
+    builtins.replaceStrings lib.upperChars (map (c: "_${c}") lib.lowerChars) s;
+
+  formatConfig = n: v:
+    let
+      formatList = x:
+        if isList x
+        then throw "can not convert 2-dimensional lists to bspwm format"
+        else formatValue x;
+
+      formatValue = v:
+        if isBool v then (if v then "true" else "false")
+        else if isList v then concatMapStringsSep ", " formatList v
+        else if isString v then "'${lib.strings.escapeShellArg v}'"
+        else toString v;
+    in
+      "bspc config ${n} ${formatValue v}";
+
+  formatMonitors = n: v: "bspc monitor ${n} -d ${concatStringsSep " " v}";
+
+  formatRules = target: directiveOptions:
+    let
+      formatDirective = n: v:
+        if isBool v then (if v then "${camelToSnake n}=on" else "${camelToSnake n}=off")
+        else if (n == "desktop" || n == "node") then "${camelToSnake n}='${v}'"
+        else "${camelToSnake n}=${lib.strings.escapeShellArg v}";
+
+      directives = filterAttrs (n: v: v != null && !(lib.strings.hasPrefix "_" n)) directiveOptions;
+      directivesStr = builtins.concatStringsSep " " (mapAttrsToList formatDirective directives);
+    in
+      "bspc rule -a ${target} ${directivesStr}";
+
+  formatStartupPrograms = map (s: "${s} &");
+
+in
+
+{
+  options = import ./options.nix { inherit pkgs; inherit lib; };
+
+  config = mkIf cfg.enable (mkMerge [
+    {
+      home.packages = [ bspwm ];
+      xsession.windowManager.command = "${cfg.package}/bin/bspwm";
+    }
+
+    (mkIf (cfg.config != null) {
+      xdg.configFile."bspwm/bspwmrc" = {
+        executable = true;
+        text = "#!/bin/sh\n\n" +
+        concatStringsSep "\n" ([]
+          ++ (optionals (cfg.monitors != {}) (mapAttrsToList formatMonitors cfg.monitors))
+          ++ [ "" ]
+          ++ (optionals (cfg.config != null) (mapAttrsToList formatConfig cfg.config))
+          ++ [ "" ]
+          ++ (optionals (cfg.rules != {}) (mapAttrsToList formatRules cfg.rules))
+          ++ [ "" ]
+          ++ (optional (cfg.applyJavaGuiFixes) ''
+            # java gui fixes
+            export _JAVA_AWT_WM_NONREPARENTING=1
```

Comment:
I would suggest removing the `applyJavaGuiFixes` option and unconditionally adding these commands to the file.